### PR TITLE
Test coverage report for C++ API (#178)

### PR DESCRIPTION
**Pull request by utilForever**
I am sorry for being late.
Add instruction to install ```lcov``` package. I had to install package on the Linux platform only because ```codecov``` needs only the results of one platform.
Please tell me whenever you find a problem. Thanks.

### CODE REVIEW

**Comment by doyubkim** [comment]
Like the title says, this seems like the package management part only. Are you planning to add actual report generation part, too? Not the codecov.io integration, but actually building Jet with covergae test (ex. https://github.com/codecov/example-cpp11-cmake/) I think this PR also should include such a build configuration, not just the dependency setup.

**Comment by utilForever** [comment]
My project does not generate the actual report file but sends it directly to ```codecov```. If you want to generate the report file, I'll work on it tomorrow.

**Comment by doyubkim** [comment]
@utilForever yes, that would be great. Please take your time, and let me know if you encounter any issues. Thanks!

**Comment by utilForever** [comment]
Sorry. I've been busy a lot due to my work last week. I'll work today.

**Comment by doyubkim** [comment]
@utilForever no problem. Again, no need to rush on this. Please take your time.

**Comment by utilForever** [comment]
I added ```CodeCoverage.cmake``` to find ```lcov``` package. Note that there are different ways to generate code coverage results via ```lcov``` when sending to Codecov and checking with a local computer. See ```.travis.yml``` and ```INSTALL.md``` for more information. Also, I set the code coverage results to appear only in Debug mode. This is because it does not produce accurate results in Release mode.

File: INSTALL.md [hunk 1]
```diff
@@ -112,6 +112,28 @@ bin\Release\unit_tests.exe
 
 There are four different tests in the codebase including the unit test, manual test, time/memory performance tests, and Python API test. For the detailed instruction on how to run those tests, please checkout the documentation page from [the project website](http://doyubkim.github.io/fluid-engine-dev/documentation/).
 
+### Code Coverage
+
+CubbyFlow-v0 uses ```lcov``` for the code coverage. For macOS and Ubuntu platforms, the code coverage report can be generated by running
```
Comment (doyubkim) [inline INSTALL.md]: CubbyFlow-v0 to Jet Framework :)

File: .travis.yml [hunk 2]
```diff
@@ -62,3 +62,21 @@ matrix:
...
+  # Creating report - Linux and debug build only
+  - if [ $TRAVIS_OS_NAME == linux ] && [ ${BUILD_TYPE} == Debug ]; then 
+       cd ${TRAVIS_BUILD_DIR};
+       sudo ln -sf /usr/bin/gcov-7 /usr/bin/gcov;
+       lcov --directory . --capture --output-file coverage.info;
+       lcov --remove coverage.info '/usr/*' --output-file coverage.info;
+       lcov --remove coverage.info '*/external/*' --output-file coverage.info;
+       lcov --remove coverage.info '*/src/tests/*' --output-file coverage.info;
+       lcov --list coverage.info;
+    fi
```
Comment (doyubkim) [inline .travis.yml]: Newline at the end of the file

File: 3RD_PARTY.md [hunk 3]
```diff
@@ -502,3 +502,7 @@ distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
+
+---
+
+Jet uses lcov for code coverage.
```
Comment (doyubkim) [inline 3RD_PARTY.md]: I don’t think we need to include lcov.

Comment (doyubkim) [reply 3RD_PARTY.md]: But we meed to include the usage of CodeCoverage.cmake

File: cmake/CompileOptions.cmake [hunk 4]
```diff
@@ -128,3 +128,19 @@ if (CMAKE_CXX_COMPILER_ID MATCHES "GNU" OR CMAKE_SYSTEM_NAME MATCHES "Linux")
         -pthread
     )
 endif()
+
+# Code coverage - Debug only
+# NOTE: Code coverage results with an optimized (non-Debug) build may be misleading
+if(CMAKE_BUILD_TYPE MATCHES Debug AND CMAKE_COMPILER_IS_GNUCXX)
```
Comment (doyubkim) [inline cmake/CompileOptions.cmake]: Space after if

File: INSTALL.md (hunk 1 above)
Comment (utilForever) [reply INSTALL.md]: Oh, it's my mistake. Fixed.

File: .travis.yml (hunk 2 above)
Comment (utilForever) [reply .travis.yml]: Fixed.

File: 3RD_PARTY.md (hunk 3 above)
Comment (utilForever) [reply 3RD_PARTY.md]: Fixed.

File: cmake/CompileOptions.cmake (hunk 4 above)
Comment (utilForever) [reply cmake/CompileOptions.cmake]: Fixed.

File: .travis.yml [hunk 5]
```diff
@@ -62,3 +62,21 @@ matrix:
...
+
+before_install:
+  - eval "${MATRIX_EVAL}"
+  - if [ $TRAVIS_OS_NAME == linux ]; then 
+      sudo apt-get install -y --no-install-recommends lcov;
+    fi
+
+after_success:
+  # Creating report - Linux and debug build only
+  - if [ $TRAVIS_OS_NAME == linux ] && [ ${BUILD_TYPE} == Debug ]; then
```
Comment (doyubkim) [inline .travis.yml]: I believe the current CMake script (`scripts/travis_build.sh`) won't really utilize Debug env var. I suggest we do not modify this `.travis.yml` file for now. I should change it anyway to integrate this repo with Codecov.io.

File: CMakeLists.txt [hunk 6]
```diff
@@ -21,29 +21,36 @@ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/bin)
...
+# NOTE: Code coverage results with an optimized (non-Debug) build may be misleading
+if(CMAKE_BUILD_TYPE MATCHES Debug AND CMAKE_COMPILER_IS_GNUCXX)
+    include(CodeCoverage)
+    setup_target_for_coverage(${PROJECT_NAME}_coverage unit_tests coverage)
 endif()
 
 # Overrides
 set(CMAKE_MACOSX_RPATH ON)
 
+# Includes
```
Comment (doyubkim) [inline CMakeLists.txt]: Any reason why the following block has moved here?

File: CMakeLists.txt [hunk 7]
```diff
@@ -21,29 +21,36 @@ set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${PROJECT_BINARY_DIR}/bin)
...
 
 # Compile options
 include(cmake/CompileOptions.cmake)
 
-# Build type - Release by default
-if (NOT CMAKE_BUILD_TYPE)
-    set(CMAKE_BUILD_TYPE Release)
+# Code coverage - Debug only
+# NOTE: Code coverage results with an optimized (non-Debug) build may be misleading
+if(CMAKE_BUILD_TYPE MATCHES Debug AND CMAKE_COMPILER_IS_GNUCXX)
```
Comment (doyubkim) [inline CMakeLists.txt]: nit: space after `if`

File: .travis.yml (hunk 5 above)
Comment (utilForever) [reply .travis.yml]: Okay. I rolled back.

File: CMakeLists.txt (hunk 7 above)
Comment (utilForever) [reply CMakeLists.txt]: Fixed.

File: CMakeLists.txt (hunk 6 above)
Comment (utilForever) [reply CMakeLists.txt]: To apply -G and -O0 flags to header files.

Comment (doyubkim) [reply CMakeLists.txt]: Hmmm but `include_directories` is just for configuring the header search path and won't consume the compile flags. Can you double check? Thanks!

Comment (utilForever) [reply CMakeLists.txt]: I did a double check and found that there was no difference in code coverage. Thank you for correcting my mistake. Fixed.

**Comment by doyubkim** [comment]
Will merge once CI completes.

**Pull request merged**